extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java

```diff
@@ -0,0 +1,95 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.incremental.oak;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.segment.incremental.AppendableIndexSpec;
+import org.apache.druid.utils.JvmUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Oak incremental index spec (describes the in-memory indexing method for data ingestion).
+ */
+public class OakIncrementalIndexSpec implements AppendableIndexSpec
+{
+  public static final String TYPE = "oak";
+
+  final long oakMaxMemoryCapacity;
+  final int oakBlockSize;
+  final int oakChunkMaxItems;
+
+  @JsonCreator
+  public OakIncrementalIndexSpec(
+      final @JsonProperty("oakMaxMemoryCapacity") @Nullable Long oakMaxMemoryCapacity,
+      final @JsonProperty("oakBlockSize") @Nullable Integer oakBlockSize,
+      final @JsonProperty("oakChunkMaxItems") @Nullable Integer oakChunkMaxItems
+  )
+  {
+    this.oakMaxMemoryCapacity = oakMaxMemoryCapacity != null && oakMaxMemoryCapacity > 0 ? oakMaxMemoryCapacity :
+        OakIncrementalIndex.Builder.DEFAULT_OAK_MAX_MEMORY_CAPACITY;
+    this.oakBlockSize = oakBlockSize != null && oakBlockSize > 0 ? oakBlockSize :
+        OakIncrementalIndex.Builder.DEFAULT_OAK_BLOCK_SIZE;
+    this.oakChunkMaxItems = oakChunkMaxItems != null && oakChunkMaxItems > 0 ? oakChunkMaxItems :
+        OakIncrementalIndex.Builder.DEFAULT_OAK_CHUNK_MAX_ITEMS;
+  }
+
+  @JsonProperty
+  public long getOakMaxMemoryCapacity()
+  {
+    return oakMaxMemoryCapacity;
+  }
+
+  @JsonProperty
+  public int getOakBlockSize()
+  {
+    return oakBlockSize;
+  }
+
+  @JsonProperty
+  public int getOakChunkMaxItems()
+  {
+    return oakChunkMaxItems;
+  }
+
+  @Nonnull
+  @Override
+  public OakIncrementalIndex.Builder builder()
+  {
+    return new OakIncrementalIndex.Builder()
+        .setOakMaxMemoryCapacity(oakMaxMemoryCapacity)
+        .setOakBlockSize(oakBlockSize)
+        .setOakChunkMaxItems(oakChunkMaxItems);
+  }
+
+  @Override
+  public long getDefaultMaxBytesInMemory()
+  {
+    // Oak allocates its keys/values directly so the JVM off-heap limitations does not apply on it.
+    // Yet, we want to respect these values if the user did not specify any specific limitation.
+    // In the realtime node, the entire JVM's direct memory is utilized for ingestion and persist operations.
+    // But maxBytesInMemory only refers to the active index size and not to the index being flushed to disk and the
+    // persist-buffer.
+    // To account for that, we set default to 1/2 of the max jvm's direct memory.
+    return JvmUtils.getRuntimeInfo().getDirectMemorySizeBytes() / 2;
```

Comment:
this will throw exception with newer versions of Java, whats a good default in that case ?

Reply:
It's a great question. In fact, I am not sure if the direct memory limit applies here at all.
Oak allocates its memory directly, so it is not subject to any JVM limitations (only to the OS memory limit).
So I'm not sure of a suitable default value.
An easy solution would be to throw an exception if this value is not set manually by the user (`maxBytesInMemory==0`).
Alternatively, rely on `maxRowsInMemory` to limit the usage and set the memory to be unlimited (same as `maxBytesInMemory==-1`).
@pjain1 @yurmix @a2l007 What do you think is the best approach here?

Reply:
@liran-funaro correct me if I am wrong when you say oak index is not subjected to any JVM limitations, isn't it subjected to Java's `XX:MaxDirectMemorySize` limit ?

Reply:
No. It allocates the memory via `Unsafe.allocateMemory()`, which is not included in the direct memory size limitation.

Reply:
So in worst case oak off heap index can allocate more than the host memory limit and OOM killer will kill the process ?

Reply:
Exactly

Reply:
In practice we set the Druid container memory limit to just a little more than Java max heap size + MaxDirectMemorySize (as per merge/query buffers config), I think its a common practice with those who run in containerized env to set both Max heap as well MaxDirectMemorySize limits to set container memory limit. 

So as per my understanding if Oak index in used in such scenario then I don't think there will be enough memory for it to work. I agree with your suggestion that we can have a middle manager config that can be set to indicate how much memory does oak index can use apart from Druid and other processes on the host. This number can be used as max bytes in memory to trigger a persist of in-memory segment to disk to prevent OOME. We should document this and say the memory limit should be more than `max heap size + MaxDirectMemorySize + oak index off heap limit + buffer for host processes`.

Reply:
A small correction, actually we do keep some amount of spare memory apart from what I have pointed out for caching segments on peons for faster querying, so Oak index can use that but will impact segment caching. Overall I think we should require setting memory limits for Oak index before enabling it as it has different memory characteristics and one will have to rethink the memory quotas allocated. @liran-funaro What do you think ?

Reply:
I agree that the admin will have to re-configure its memory quotas when enabling this extension.

The on-heap index uses `Xmx/6` as a limitation (it uses `Xmx/3` overall because we can have two "live" indexes).
Since Oak doesn't need that heap space, the admin can set `Xms` to be `(2/3)*Xmx`, then the default Oak usage will be the same as the on-heap one.
I assume here that the JVM won't use more memory than needed (i.e., `Xms`).

So the extension's docs should suggest setting `Xms=(2/3)*Xmx`. @pjain1 @yurmix What do you think?

Reply:
Putting it in another way the docs can say that oak index is off heap so Xmx can be now be set to `2/3` of the earlier value. To keep memory requirements same please set max bytes in memory to `1/6` of Xmx ? 

How about `getDefaultMaxBytesInMemory` here return `JvmUtils.getRuntimeInfo().getMaxHeapSizeBytes() / 6` same as for `OnheapIncrementalIndex` so that default memory requirement is same as earlier when Xmx is set to `2/3` of earlier value ?

Reply:
The problem with changing `Xmx` is that the ingestion task might be configured to use the on-heap index. Then it will have less memory for the index.
I think it is OK to rely on `Xms` because the Peon task has very little use of the heap.

Reply:
Ok this makes sense and keeps config backward compatible, but coming back to the point of this thread -
1. What should be the default value returned from `getDefaultMaxBytesInMemory` ? I think `JvmUtils.getRuntimeInfo().getMaxHeapSizeBytes() / 6` makes sense when Xms is set as per your recommendation.
2.  Docs should be make clear about oak memory characteristics and how to set host memory limits if `maxBytesInMemory` is set by user so that tasks do not fail with OOME.

Reply:
Agreed! I'll update the default value and the docs accordingly. Thanks!